fw/http2.c

```diff
@@ -0,0 +1,583 @@
+
+        return tfw_hpack_init(&ctx->hpack, HPACK_TABLE_DEF_SIZE);
+}
+
+void
+tfw_h2_context_clear(TfwH2Ctx *ctx)
+{
+        WARN_ON_ONCE(ctx->streams_num);
+        tfw_hpack_clean(&ctx->hpack);
+}
+
+void
+tfw_h2_conn_terminate_close(TfwH2Ctx *ctx, TfwH2Err err_code, bool close,
+                            bool attack)
+{
+        TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+
+        if (tfw_h2_send_goaway(ctx, err_code, attack) && close) {
+                if (attack)
+                        tfw_connection_close((TfwConn *)conn, true);
+                else
+                        tfw_connection_shutdown((TfwConn *)conn, true);
+        }
+}
+
+void
+tfw_h2_conn_streams_cleanup(TfwH2Ctx *ctx)
+{
+        TfwStream *cur, *next;
+        TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+        TfwStreamSched *sched = &ctx->sched;
+
+        WARN_ON_ONCE(((TfwConn *)conn)->stream.msg);
+
+        T_DBG3("%s: ctx [%p] conn %p sched %p\n", __func__, ctx, conn, sched);
+
+        tfw_h2_remove_idle_streams(ctx, UINT_MAX);
+
+        rbtree_postorder_for_each_entry_safe(cur, next, &sched->streams, node) {
+                tfw_h2_stream_purge_all_and_free_response(cur);
+                tfw_h2_stream_unlink_lock(ctx, cur);
+
+                /* The streams tree is about to be destroyed and
+                 * we don't want to trigger rebalancing.
+                 * No further actions regarding streams dependencies/prio
+                 * is required at this stage.
+                 */
+                tfw_h2_delete_stream(cur);
+                --ctx->streams_num;
+        }
+        sched->streams = RB_ROOT;
+}
+
+void
+tfw_h2_current_stream_remove(TfwH2Ctx *ctx)
+{
+        T_DBG3("%s: ctx [%p] ctx->cur_stream %p\n", __func__,
+               ctx, ctx->cur_stream);
+        tfw_h2_stream_unlink_lock(ctx, ctx->cur_stream);
+        tfw_h2_stream_clean(ctx, ctx->cur_stream);
+        ctx->cur_stream = NULL;
+}
+
+/**
+ * According to RFC 9113 section 5.1.1:
+ * The first use of a new stream identifier implicitly closes all
+ * streams in the "idle" state that might have been initiated by that
+ * peer with a lower-valued stream identifier.
+ */
+void
+tfw_h2_remove_idle_streams(TfwH2Ctx *ctx, unsigned int id)
+{
+        TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+        TfwStream *stream, *tmp;
+
+        /*
+         * We add and remove streams from idle queue under
+         * socket lock.
+         */
+        assert_spin_locked(&((TfwConn *)conn)->sk->sk_lock.slock);
+
+        list_for_each_entry_safe_reverse(stream, tmp, &ctx->idle_streams.list,
+                                         hcl_node)
+        {
+                if (id <= stream->id)
+                        break;
+
+                tfw_h2_stream_del_from_queue_nolock(stream);
+                tfw_h2_set_stream_state(stream, HTTP2_STREAM_CLOSED);
+                tfw_h2_stream_add_closed(ctx, stream);
+        }
+}
```

Comment:
This function is used above, so let's move it before `tfw_h2_conn_streams_cleanup()`.

Reply:
Yes fixed